zkevm-circuits/src/super_circuit.rs

```diff
@@ -336,6 +341,67 @@ impl<const MAX_TXS: usize, const MAX_CALLDATA: usize> SuperCircuit<Fr, MAX_TXS,
         };
         Ok((k, circuit, instance))
     }
+
+    /// From the witness data, generate a SuperCircuit instance with all of the
+    /// sub-circuits filled with their corresponding witnesses.
+    ///
+    /// Also, return with it the minimum required SRS degree for the circuit and
+    /// the Public Inputs needed.
+    pub fn build_with_builder(
+        builder: CircuitInputBuilder,
+        eth_block: eth_types::Block<eth_types::Transaction>,
+        rng: &mut (impl RngCore + Clone),
+    ) -> Result<(u32, Self, Vec<Vec<Fr>>), bus_mapping::Error> {
+        let keccak_inputs = builder.keccak_inputs()?;
+        let mut block = block_convert(&builder.block, &builder.code_db);
+
+        block.randomness = Fr::random(rng.clone());
+        let aux_generator = <Secp256k1Affine as CurveAffine>::CurveExt::random(rng).to_affine();
+
+        let fixed_table_tags: Vec<FixedTableTag> = FixedTableTag::iter().collect();
+        let log2_ceil = |n| u32::BITS - (n as u32).leading_zeros() - (n & (n - 1) == 0) as u32;
+
+        let num_rows_required_for_steps =
+            SuperCircuit::<_, MAX_TXS, MAX_CALLDATA>::get_num_rows_required(&block);
+
+        let k = log2_ceil(
+            64 + fixed_table_tags
+                .iter()
+                .map(|tag| tag.build::<Fr>().count())
+                .sum::<usize>(),
+        );
+        let bytecodes_len = block
+            .bytecodes
+            .iter()
+            .map(|(_, bytecode)| bytecode.bytes.len())
+            .sum::<usize>();
+        let k = k.max(log2_ceil(64 + bytecodes_len));
+        let k = k.max(log2_ceil(64 + num_rows_required_for_steps));
+        let k = k + 1;
```

Comment:
It seems we need to increate k value. Otherwise, we will get row not enough.

Reply:
So the calculation of `k` is currently not accurate enough to correctly calculate the circuit size? Is that also true for the calculations done elsewhere to calculate the height?

Reply:
are those columns in different circuits/gadgets independent (and unrelated excepts copy constraints)? If so, I guess here max missed few more circuits as it only takes bytecode_len (for bytecode_circuit) & num_rows_._steps (for evm_circuit I think) into account. 
I think it depends on the layouter, so ask halo2 for the mininum k (if possible) is better than counting by the programmer oneself.

Reply:
The columns used for the lookups are shared. But yeah it doesn't seem to take some circuits into account. I think the layouter shouldn't have an impact because all columns are simply put next to each other, but not completely sure.

I don't think there's a good way to let halo2 calculate the height because halo2 has limited knowledge about the circuits, the only think it can do is complain if there aren't enough rows (which you could use to make the circuit bigger and try again but that's very inefficient).

The super circuit has a function to calculate the number of rows, but it seems like it's currently not really used (only calls into the EVM circuit that exposes the same function): https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/673f5fe6e9e76ce3796a3f5b1d15d7c49ced66cd/zkevm-circuits/src/super_circuit.rs#L127

I think that would be a good place to do these calculations and the same type function for circuits that don't yet have so each circuit can calculate on it's own how many rows are needed for a specific witness.

Reply:
> I don't think there's a good way to let halo2 calculate the height because halo2 has limited knowledge about the circuits, the only think it can do is complain if there aren't enough rows (which you could use to make the circuit bigger and try again but that's very inefficient).

Yes, you're right, The counting actually ask halo2 to do sort of 2-pass analysis of the whole circuit as the request row number is decided only after synthesize() done, in the first round it does nothing excepts counting & report, or maybe halo2 could have a kind of allocator to enlarge itself dynamically, but we should not rely on (expect to) that so far.
BTW: I saw an issue on dup witness generation of halo2 https://github.com/taikochain/zkevm-circuits/issues/7. I just feel ( bad feeling? ) they are similar issue , maybe we can also makes halo2 smart enough to know the required rows together.

> The super circuit has a function to calculate the number of rows, but it seems like it's currently not really used (only calls into the EVM circuit that exposes the same function)

I think the estimation of k could be more accurate if all circuits contained this API. However, maybe estimating k is not important enough (4~5 tries for programmer only) to take extra coding efforts.

Reply:
> we can also makes halo2 smart enough to know the required rows together.

There is still a hard limit on how big you want to make the circuit (trusted setup limitation and what not), so you don't really want halo2 to keep making the circuit bigger in all cases. And if there's some cap, you still get in the same situation. Also in this case we actually want to fill up the circuit up to some size, not by chance create some circuit that is only filled a bit more than half.

> I saw an issue on dup witness generation of halo2 https://github.com/taikochain/zkevm-circuits/issues/7. I just feel ( bad feeling? ) they are similar issue.

I think they are a bit different because (as far as I know) the layouter will never change the height of the circuit, just lay out stuff within the circuit differently.

> I think the estimation of k could be more accurate if all circuits contained this API. However, maybe estimating k is not important enough (4~5 tries for programmer only) to take extra coding efforts.

I think it's kind of nice. In most cases you actually want to know how big you have to make the circuit do a certain thing, so it's even nice just to have the formula there in the code instead of having to figure out how the full circuit works.